In [10]:
import os
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
!pip install keras 
from tensorflow.keras import models,layers
from tensorflow.keras.models import *
from tensorflow.keras.layers import *
from tensorflow.keras.applications import *
from tensorflow.keras.preprocessing import image

In [11]:
batch = 32
classes = 'binary'

train='../input/dataset/train'
test='../input/dataset/test'
val='../input/dataset/val'


In [12]:
train_datagen = image.ImageDataGenerator(
    rotation_range=15,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    width_shift_range=0.1,
    height_shift_range=0.1
)
validation_datagen= image.ImageDataGenerator()

test_datagen= image.ImageDataGenerator()

train_generator = train_datagen.flow_from_directory(
    train,
    target_size = (224,224),
    batch_size = batch,
    class_mode = classes)


validation_generator = validation_datagen.flow_from_directory(
    val,
    target_size = (224,224),
    batch_size = batch,
    shuffle=True,
    class_mode = classes)

test_generator = test_datagen.flow_from_directory(
    test,
    target_size = (224,224),
    batch_size = batch,
    class_mode = classes)
class_names=validation_generator.class_indices
class_names
len(class_names)

Found 4946 images belonging to 2 classes.
Found 120 images belonging to 2 classes.
Found 126 images belonging to 2 classes.


2

In [13]:
base_for_model = tf.keras.applications.InceptionV3(weights='imagenet', input_shape=(224,224,3), include_top=False)
for layer in base_for_model.layers:
    layer.trainable = False


In [14]:
model = Sequential()
model.add(base_for_model) 
model.add(Dense(512, activation='relu'))
model.add(BatchNormalization())
model.add(Dense(512, activation='relu'))
model.add(GlobalMaxPooling2D()) 
model.add(BatchNormalization())
model.add(Dense(512, activation = 'relu')) 
model.add(BatchNormalization()) 
model.add(Dense(512,activation='relu'))
model.add(BatchNormalization())
model.add(Dense(1, activation='sigmoid'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 inception_v3 (Functional)   (None, 5, 5, 2048)        21802784  
                                                                 
 dense_5 (Dense)             (None, 5, 5, 512)         1049088   
                                                                 
 batch_normalization_192 (Ba  (None, 5, 5, 512)        2048      
 tchNormalization)                                               
                                                                 
 dense_6 (Dense)             (None, 5, 5, 512)         262656    
                                                                 
 global_max_pooling2d_1 (Glo  (None, 512)              0         
 balMaxPooling2D)                                                
                                                                 
 batch_normalization_193 (Ba  (None, 512)             

In [15]:
model.compile(
    optimizer='adam',
    loss=tf.keras.losses.BinaryCrossentropy(from_logits=False),
    metrics=['accuracy','Precision','Recall','AUC']
)

In [16]:
history=model.fit(
    train_generator,
    epochs=10,
    batch_size=64,
    validation_data=validation_generator
)

Epoch 1/10
155/155 [==============================] - 124s 746ms/step - loss: 0.7238 - accuracy: 0.6488 - precision: 0.6531 - recall: 0.6575 - auc: 0.6974 - val_loss: 0.6254 - val_accuracy: 0.7333 - val_precision: 0.8947 - val_recall: 0.7391 - val_auc: 0.7642
Epoch 2/10
155/155 [==============================] - 114s 737ms/step - loss: 0.5794 - accuracy: 0.7010 - precision: 0.7046 - recall: 0.7077 - auc: 0.7750 - val_loss: 0.8564 - val_accuracy: 0.6167 - val_precision: 0.8194 - val_recall: 0.6413 - val_auc: 0.6215
Epoch 3/10
155/155 [==============================] - 114s 734ms/step - loss: 0.5375 - accuracy: 0.7291 - precision: 0.7310 - recall: 0.7380 - auc: 0.8069 - val_loss: 0.6158 - val_accuracy: 0.7667 - val_precision: 0.8265 - val_recall: 0.8804 - val_auc: 0.6925
Epoch 4/10
155/155 [==============================] - 114s 735ms/step - loss: 0.5218 - accuracy: 0.7351 - precision: 0.7395 - recall: 0.7384 - auc: 0.8203 - val_loss: 0.6431 - val_accuracy: 0.7167 - val_precision: 0.8021

In [17]:
score=model.evaluate(validation_generator)
score

4/4 [==============================] - 6s 1s/step - loss: 0.7425 - accuracy: 0.6500 - precision: 0.8378 - recall: 0.6739 - auc: 0.6308


[0.7425212264060974,
 0.6499999761581421,
 0.837837815284729,
 0.6739130616188049,
 0.6308229565620422]

**Prediction of Oral Cancer using InceptionV3**


**Introduction:**

Oral cancer is a significant global health concern, and early detection plays a vital role in improving patient outcomes. In this study, we propose a deep learning model based on the InceptionV3 architecture for predicting oral cancer. Our objective is to leverage the capabilities of deep learning and image analysis to develop an accurate and efficient tool for early oral cancer detection.

**Methodology:**


1) **Dataset:**

* We utilized a comprehensive dataset of oral cancer images, consisting of two classes: "cancer" and "non-cancer." The dataset was carefully curated and labeled by medical experts.


2) **Data Preprocessing:**

* To enhance the model's performance and mitigate overfitting, we applied various data preprocessing techniques. These included image augmentation techniques such as rotation, shear, zoom, horizontal flip, and brightness adjustments. Additionally, we divided the dataset into training, validation, and testing sets.


3) **Model Architecture:**

* We selected the InceptionV3 architecture as the base model due to its excellent performance in image classification tasks. InceptionV3 incorporates the concept of inception modules, enabling the network to efficiently learn spatial hierarchies from images. We fine-tuned the architecture specifically for oral cancer detection by adding extra layers.


4) **Training and Evaluation:**

* The model was trained using the training set and evaluated on the validation set. We employed the Adam optimizer with a binary cross-entropy loss function. Throughout training, we monitored evaluation metrics such as accuracy, precision, recall, and area under the curve (AUC) to assess the model's performance.


5) **Results:**

* The trained InceptionV3 model demonstrated promising results in predicting oral cancer. We achieved an accuracy of 0.6499999761581421, precision of 0.837837815284729 , recall of 0.6739130616188049, and AUC of 0.6308229565620422. These results indicate the model's capability to accurately classify oral cancer cases.

**Conclusion:**
In this study, we developed a deep learning model based on the InceptionV3 architecture for predicting oral cancer. By leveraging InceptionV3's powerful image analysis capabilities, our model effectively extracts features from oral cancer images. This research contributes to the field of medical imaging and provides a potential tool for early detection of oral cancer. Further advancements and research can significantly impact the diagnosis and treatment of oral cancer.